altos-rust/altos-core/src/sched/mod.rs

```diff
@@ -74,26 +80,52 @@ pub fn switch_context() {
         }
       }
 
-      'main: loop {
-        for i in Priority::all() {
-          while let Some(mut new_task) = PRIORITY_QUEUES[i].dequeue() {
-            if new_task.destroy {
-              drop(new_task);
-            }
-            else {
-              new_task.state = State::Running;
-              // UNSAFE: Accessing CURRENT_TASK
-              unsafe { CURRENT_TASK = Some(new_task) };
-              break 'main;
-            }
-          }
-        }
-      }
+      //NOTE May cause additional overhead by calling a stack frame
+      while schedule_next_task() { }
     },
     None => panic!("switch_context - current task doesn't exist!"),
   }
 }
 
+///Get the next task from the queue
+///If more than NORMAL_TASK_MAX Normal tasks have run don't try and schedule
+///a normal priorty task and give low priority a shot at running.
+///NOTE: Ranges could be replaced with constants to improve performace
+fn schedule_next_task() -> bool {
+  if NORMAL_TASK_COUNTER.load(Ordering::Relaxed)  >= NORMAL_TASK_MAX {
+    if !select_task(Priority::get_range_from(Priority::Critical, Priority::Normal)) {
```

Comment:
This range only includes critical, right? At first glance, it looks like it would include critical and normal, and I think the code would be incorrect if it included normal since it would still pick that over low priority tasks. Also, there's an extra space on line 95 right before >=.